x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -490,7 +492,87 @@ static XContentBuilder buildUpdatedDocument(
+        final UpdateApiKeyRequest request,
+        final Set<RoleDescriptor> userRoles
+    ) {
+        if (apiKeyDoc.version != targetDocVersion.id) {
+            return false;
+        }
+
+        final Map<String, Object> currentCreator = apiKeyDoc.creator;
+        final var user = authentication.getEffectiveSubject().getUser();
+        final var sourceRealm = authentication.getEffectiveSubject().getRealm();
+        if (false == (Objects.equals(user.principal(), currentCreator.get("principal"))
+            && Objects.equals(user.fullName(), currentCreator.get("full_name"))
+            && Objects.equals(user.email(), currentCreator.get("email"))
+            && Objects.equals(user.metadata(), currentCreator.get("metadata"))
+            && Objects.equals(sourceRealm.getName(), currentCreator.get("realm"))
+            && Objects.equals(sourceRealm.getType(), currentCreator.get("realm_type")))) {
+            return false;
+        }
+        if (sourceRealm.getDomain() != null) {
+            if (currentCreator.get("realm_domain") == null) {
+                return false;
+            }
+            @SuppressWarnings("unchecked")
+            final var currentRealmDomain = RealmDomain.fromXContent(
+                XContentHelper.mapToXContentParser(
+                    XContentParserConfiguration.EMPTY,
+                    (Map<String, Object>) currentCreator.get("realm_domain")
+                )
+            );
+            if (sourceRealm.getDomain().equals(currentRealmDomain) == false) {
+                return false;
+            }
+        } else {
+            if (currentCreator.get("realm_domain") != null) {
+                return false;
+            }
+        }
+
+        final Map<String, Object> newMetadata = request.getMetadata();
+        if (newMetadata != null) {
+            if (apiKeyDoc.metadataFlattened == null) {
+                return false;
+            }
+            final Map<String, Object> currentMetadata = XContentHelper.convertToMap(apiKeyDoc.metadataFlattened, false, XContentType.JSON)
+                .v2();
+            if (newMetadata.equals(currentMetadata) == false) {
+                return false;
+            }
+        }
+
+        final List<RoleDescriptor> newRoleDescriptors = request.getRoleDescriptors();
+        if (newRoleDescriptors != null) {
+            final List<RoleDescriptor> currentRoleDescriptors = parseRoleDescriptorsBytes(
+                request.getId(),
+                apiKeyDoc.roleDescriptorsBytes,
+                RoleReference.ApiKeyRoleType.ASSIGNED
+            );
+            if (false == (newRoleDescriptors.size() == currentRoleDescriptors.size()
+                && new HashSet<>(newRoleDescriptors).equals(new HashSet<>(currentRoleDescriptors)))) {
+                return false;
+            }
+        }
+
+        assert userRoles != null;
+        final List<RoleDescriptor> currentLimitedByRoleDescriptorRoles = parseRoleDescriptorsBytes(
```

Comment:
Nit: The name `currentLimitedByRoleDescriptorRoles` is a bit awkward. It should just be `currentLimitedByRoleDescriptors`. Also, `userRoles` really should be `userRoleDescriptors` since the code has distinct concepts for role, role name and role descriptor.

Reply:
Oh whoops `currentLimitedByRoleDescriptorRoles` is definitely a copy&mishap. Thanks for catching 👍 

> Also, userRoles really should be userRoleDescriptors since the code has distinct concepts for role, role name and role descriptor.

I'm good with that, I went with `userRoles` because that's what it was called before the work on updatable API keys elsewhere in this class.

Reply:
> I went with userRoles because that's what it was called before the work on updatable API keys elsewhere in this class.

I figured. I wasn't a fan of this name in other places as well. But we never bothered to change them ...